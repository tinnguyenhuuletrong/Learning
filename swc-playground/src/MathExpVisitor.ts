import {
  AssignmentExpression,
  BinaryExpression,
  Expression,
  ExpressionStatement,
  NumericLiteral,
} from "@swc/core/types";
import { Visitor } from "@swc/core/Visitor";

export class RuntimeData {
  stack = [];
  mem = {};
  _logs = [];

  _val = new WeakMap();
}

export class RuntimeError extends Error {
  ctx?: RuntimeData = null;
  extraInfo? = null;

  constructor(
    message: string,
    { ctx, extraInfo }: { ctx: RuntimeData; extraInfo?: any }
  ) {
    super(message);
    this.name = "RuntimeError";

    this.ctx = ctx;
    this.extraInfo = extraInfo;
  }
}

function asNumber(ctx: RuntimeData, exp: Expression): number {
  switch (exp.type) {
    case "NumericLiteral":
      return exp.value;
    case "Identifier":
      return ctx.mem[exp.value];
    case "ParenthesisExpression": {
      const val = ctx._val.get(exp.expression);
      if (val) return val;

      throw new RuntimeError(`asNumber failed ${exp.type}`, { ctx });
    }
    default: {
      const val = ctx._val.get(exp);
      if (val) return val;

      throw new RuntimeError(`asNumber failed ${exp.type}`, { ctx });
    }
  }
}

function asIdentifier(ctx: RuntimeData, exp: Expression) {
  switch (exp.type) {
    case "StringLiteral":
      return exp.value;
    case "Identifier":
      return exp.value;
  }
}

function asValue(ctx: RuntimeData, exp: Expression) {
  const val = ctx._val.get(exp);
  return val;
}

export class MathExpVisitor extends Visitor {
  ctx: RuntimeData;

  constructor() {
    super();
    this.ctx = new RuntimeData();
  }

  visitAssignmentExpression(n: AssignmentExpression) {
    super.visitAssignmentExpression(n);
    const varName = asIdentifier(this.ctx, n.left as Expression);
    const varVal = asValue(this.ctx, n.right);
    this.ctx.mem[varName] = varVal;
    this.ctx._logs.push(`assign ${varName} = ${varVal}`);
    return n;
  }

  visitExpressionStatement(n: ExpressionStatement) {
    super.visitExpressionStatement(n);
    this.ctx._val.set(n, n.expression);
    return n;
  }

  visitBinaryExpression(n: BinaryExpression) {
    super.visitBinaryExpression(n);
    const op = n.operator;
    // can left / right be NumericLiteral / Expression / Identify
    const v1 = asNumber(this.ctx, n.left);
    const v2 = asNumber(this.ctx, n.right);
    let res;
    switch (op) {
      case "+":
        res = v1 + v2;
        break;
      case "-":
        res = v1 - v2;
        break;
      case "*":
        res = v1 * v2;
        break;
      case "/":
        res = v1 / v2;
        break;
      case "%":
        res = v1 % v2;
        break;
      case "**":
        res = v1 ** v2;
        break;
      case "==":
        res = v1 == v2;
        break;
      default:
        throw new Error(`unknown op ${op}`);
    }
    this.ctx._val.set(n, res);
    this.ctx._logs.push(`exec ${op} for r:${v1}, l:${v2} -> ${res}`);
    return n;
  }

  visitNumericLiteral(n: NumericLiteral) {
    const v = parseFloat(n.raw);
    this.ctx._val.set(n, v);
    this.ctx._logs.push(`got num: ${v}`);
    return n;
  }
}
