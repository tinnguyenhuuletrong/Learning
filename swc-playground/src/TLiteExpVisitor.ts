import {
  AssignmentExpression,
  BinaryExpression,
  BooleanLiteral,
  ConditionalExpression,
  Expression,
  ExpressionStatement,
  IfStatement,
  Module,
  NumericLiteral,
  ParenthesisExpression,
  Statement,
  StringLiteral,
  UnaryExpression,
} from "@swc/core/types";
import { Visitor } from "@swc/core/Visitor";

declare module "@swc/core/types" {
  interface Node {
    _runtimeValue: any;
  }
  interface Fn {
    _runtimeValue: Function;
  }
}

type RuntimeContextOption = {
  debugTrace: boolean;
};

export class RuntimeContext {
  stack = [];
  mem = {};
  _logs = [];

  private _opts: RuntimeContextOption;

  constructor(opts: RuntimeContextOption = { debugTrace: true }) {
    this._opts = opts;
  }

  addLog(inp: string) {
    if (this._opts.debugTrace) this._logs.push(inp);
  }
}

export class RuntimeError extends Error {
  ctx?: RuntimeContext = null;
  extraInfo? = null;

  constructor(
    message: string,
    { ctx, extraInfo }: { ctx: RuntimeContext; extraInfo?: any }
  ) {
    super(message);
    this.name = "RuntimeError";

    this.ctx = ctx;
    this.extraInfo = extraInfo;
  }
}

function asIdentifier(ctx: RuntimeContext, exp: Expression) {
  switch (exp.type) {
    case "StringLiteral":
      return exp.value;
    case "Identifier":
      return exp.value;
  }
}

function resolveValue(ctx: RuntimeContext, n: Expression) {
  switch (n.type) {
    case "Identifier":
      return ctx.mem[n.value];

    default:
      return n._runtimeValue;
  }
}

export class TLiteExpVisitor extends Visitor {
  ctx: RuntimeContext;

  constructor() {
    super();
  }

  run(astCode: Module, ctx: RuntimeContext) {
    this.ctx = ctx;
    const res = this.visitProgram(astCode);
    return res;
  }

  visitAssignmentExpression(n: AssignmentExpression) {
    super.visitAssignmentExpression(n);
    const varName = asIdentifier(this.ctx, n.left as Expression);
    const varVal = n.right._runtimeValue;
    this.ctx.mem[varName] = varVal;
    this.ctx.addLog(`assign ${varName} = ${varVal}`);
    return n;
  }

  visitParenthesisExpression(n: ParenthesisExpression) {
    super.visitParenthesisExpression(n);
    n._runtimeValue = n.expression._runtimeValue;
    return n;
  }

  visitUnaryExpression(n: UnaryExpression): Expression {
    super.visitUnaryExpression(n);

    const op = n.operator;
    const v = resolveValue(this.ctx, n.argument);
    let res;
    switch (op) {
      case "!":
        res = !v;
        break;
      case "+":
        res = +v;
        break;
      case "-":
        res = -v;
        break;
      case "~":
        res = ~v;
        break;

      default:
        throw new Error(`UnaryExpression: unknown op ${op}`);
    }

    n._runtimeValue = res;
    this.ctx.addLog(`exec ${op} for v:${v} -> ${res}`);
    return n;
  }

  visitBinaryExpression(n: BinaryExpression) {
    super.visitBinaryExpression(n);
    const op = n.operator;
    // can left / right be NumericLiteral / Expression / Identify
    const v1 = resolveValue(this.ctx, n.left);
    const v2 = resolveValue(this.ctx, n.right);
    let res;
    switch (op) {
      case "+":
        res = v1 + v2;
        break;
      case "-":
        res = v1 - v2;
        break;
      case "*":
        res = v1 * v2;
        break;
      case "/":
        res = v1 / v2;
        break;
      case "%":
        res = v1 % v2;
        break;
      case "**":
        res = v1 ** v2;
        break;

      case "===":
      case "==":
        res = v1 == v2;
        break;

      case "!=":
      case "!==":
        res = v1 != v2;
        break;

      case ">":
        res = v1 > v2;
        break;
      case ">=":
        res = v1 >= v2;
        break;

      case "<":
        res = v1 < v2;
        break;
      case "<=":
        res = v1 <= v2;
        break;

      case "&&":
        res = v1 && v2;
        break;
      case "||":
        res = v1 || v2;
        break;

      default:
        throw new Error(`unknown op ${op}`);
    }
    n._runtimeValue = res;
    this.ctx.addLog(`exec ${op} for v1:${v1}, v2:${v2} -> ${res}`);
    return n;
  }

  visitConditionalExpression(n: ConditionalExpression): Expression {
    this.ctx.addLog(`exec if_branch_expression test`);
    n.test = this.visitExpression(n.test);
    const condVal = n.test._runtimeValue;
    let res;

    if (condVal) {
      n.consequent = this.visitExpression(n.consequent);
      res = n.consequent._runtimeValue;
    } else {
      n.alternate = this.visitExpression(n.alternate);
      res = n.alternate._runtimeValue;
    }

    n._runtimeValue = res;
    this.ctx.addLog(`exec if_branch_expression: ${condVal} -> ${res}`);
    return n;
  }

  visitIfStatement(n: IfStatement) {
    this.ctx.addLog(`exec if_branch test`);
    n.test = this.visitExpression(n.test);
    const condVal = n.test._runtimeValue;

    if (condVal) {
      n.consequent = this.visitStatement(n.consequent);
    } else {
      n.alternate = this.visitOptionalStatement(n.alternate);
    }

    this.ctx.addLog(`exec if_branch: ${condVal}`);
    return n;
  }

  visitNumericLiteral(n: NumericLiteral) {
    const v = parseFloat(n.raw);
    this.ctx.addLog(`got num: ${v}`);
    n._runtimeValue = v;
    return n;
  }

  visitStringLiteral(n: StringLiteral) {
    const v = n.value;
    this.ctx.addLog(`got str: ${v}`);
    n._runtimeValue = v;
    return n;
  }

  visitBooleanLiteral(n: BooleanLiteral) {
    const v = n.value;
    this.ctx.addLog(`got bool: ${v}`);
    n._runtimeValue = v;
    return n;
  }
}
