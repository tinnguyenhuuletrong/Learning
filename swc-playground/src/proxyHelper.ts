export const wrappedWithFuncCallLog = <T = any>(
  ins: T,
  { withFuncParamLog }: { withFuncParamLog: boolean } = {
    withFuncParamLog: false,
  }
): T => {
  return new Proxy(ins as any, {
    get: function (target, name, receiver) {
      if (!target.hasOwnProperty(name)) {
        if (
          typeof target[name] === "function" &&
          String(name).includes("visit")
        ) {
          console.log("Calling method:", name);
        }
        return function () {
          if (withFuncParamLog)
            console.log(`\tparams: ${JSON.stringify(arguments)}`);
          return target[name].apply(this, arguments);
        };
      }
      return Reflect.get(target, name, receiver);
    },
  }) as unknown as T;
};
