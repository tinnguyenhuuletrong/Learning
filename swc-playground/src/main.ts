import * as swc from "@swc/core";
import { TLiteJITEngine, RuntimeContext } from "./TLiteJITEngine";
import { wrappedWithFuncCallLog } from "./proxyHelper";
import { isArray, isFunction } from "lodash";

// https://astexplorer.net/

const code = `
function add(a, b) {
  if(a < 10) return 1
  return a + b
}
res = add(1,2)
`;

const res = swc.parseSync(code);
const ins = wrappedWithFuncCallLog(new TLiteJITEngine());
const ctx = new RuntimeContext();
ctx.mem = { inp: 2 };
ctx.funcDb["map"] = (inp, func) => {
  if (!isFunction(func)) throw new Error("arg1 is not a function");
  if (!isArray(inp)) throw new Error("arg0 is not a array");
  return inp.map((itm) => func(itm));
};

const finalVal = ins.run(res, ctx);
console.dir(
  {
    ctx: ctx,
  },
  { depth: 5 }
);
