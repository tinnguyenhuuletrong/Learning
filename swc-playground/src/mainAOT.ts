import * as swc from "@swc/core";
import {
  TLiteAotCompileVisitor,
  CompilerContext,
} from "./TLiteAotCompileVisitor";
import { wrappedWithFuncCallLog } from "./proxyHelper";
import { isArray, isFunction } from "lodash";
import {
  RuntimeAotContext,
  RuntimeAotError,
  TLiteAotEngine,
} from "./TLiteAotEngine";
import { Op } from "./type.aot";

// https://astexplorer.net/
const code = `
if(inp > 2)
  res.status = true
else {
  res.status = false
  res.extra = "reason"
}
res.tmp = inp > 100 ? '1' : '0'
`;

const ops = compile(code);
run(clone(ops));

function compile(code: string) {
  const res = swc.parseSync(code);
  const ins = wrappedWithFuncCallLog(new TLiteAotCompileVisitor());
  const ctx = new CompilerContext();

  const finalVal = ins.run(res, ctx);
  console.dir(
    {
      ctx: ctx,
    },
    { depth: 20 }
  );

  console.log(ctx.toString());
  console.log(code);
  return ctx.ops;
}

function run(ops: Op[]) {
  const runtime = new TLiteAotEngine();
  const ctx = new RuntimeAotContext({ debugTrace: true });
  ctx.mem["inp"] = -999;
  const res = runtime.run(ops, ctx);

  console.log(res);
  console.dir(
    {
      mem: ctx.mem,
      logs: ctx._logs,
    },
    { depth: 20 }
  );
}

function clone(obj: any) {
  return JSON.parse(JSON.stringify(obj));
}
