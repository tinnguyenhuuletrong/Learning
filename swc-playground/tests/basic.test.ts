import * as swc from "@swc/core";
import { TLiteExpVisitor, RuntimeContext } from "../src/TLiteExpVisitor";

describe("TLite expression", () => {
  test("1 + (3 * 5)**2", async () => {
    const exp = `1 + (3 * 5)**2;`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();

    const finalVal = runtime.run(astTree, new RuntimeContext());
    expect(
      finalVal.body[0].type === "ExpressionStatement" &&
        finalVal.body[0].expression._runtimeValue === 226
    ).toBe(true);
    expect(runtime).toMatchSnapshot();
  });

  test("res = 1 + (3 * 5)**2", async () => {
    const exp = `res = 1 + (3 * 5)/2;`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();

    runtime.run(astTree, new RuntimeContext());
    expect(runtime.ctx.mem["res"]).toBe(8.5);
    expect(runtime).toMatchSnapshot();
  });

  test("res = inp > 2", async () => {
    const exp = `res = inp > 2`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();
    const ctx = new RuntimeContext();
    ctx.mem["inp"] = 3;
    runtime.run(astTree, ctx);
    expect(runtime.ctx.mem["res"]).toBe(true);
    expect(runtime).toMatchSnapshot();
  });

  test("res = inp <= 2", async () => {
    const exp = `res = inp <= 2`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();
    const ctx = new RuntimeContext();
    ctx.mem["inp"] = 2;
    runtime.run(astTree, ctx);
    expect(runtime.ctx.mem["res"]).toBe(true);
    expect(runtime).toMatchSnapshot();
  });

  test("res = inp > 2 && inp <= 5", async () => {
    const exp = `res = inp > 2 && inp <= 5`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();
    const ctx = new RuntimeContext();
    ctx.mem["inp"] = 10;
    runtime.run(astTree, ctx);
    expect(runtime.ctx.mem["res"]).toBe(false);
    expect(runtime).toMatchSnapshot();
  });

  test("res = inp < 2 || inp > 5", async () => {
    const exp = `res = inp < 2 || inp > 5`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();
    const ctx = new RuntimeContext();
    ctx.mem["inp"] = -1;
    runtime.run(astTree, ctx);
    expect(runtime.ctx.mem["res"]).toBe(true);
    expect(runtime).toMatchSnapshot();
  });

  test("res = !(+inp < 2 || +inp > 5)", async () => {
    const exp = `res = !(+inp < 2 || +inp > 5)`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();
    const ctx = new RuntimeContext();
    ctx.mem["inp"] = "-1";
    runtime.run(astTree, ctx);
    expect(runtime.ctx.mem["res"]).toBe(false);
    expect(runtime).toMatchSnapshot();
  });
});

describe("TLite condition", () => {
  test("res = inp > 18 ? 'mature' : 'teen'", async () => {
    const exp = `res = inp > 18 ? 'mature' : 'teen'`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();
    const ctx = new RuntimeContext();
    ctx.mem["inp"] = "22";
    runtime.run(astTree, ctx);
    expect(runtime.ctx.mem["res"]).toBe("mature");
    expect(runtime).toMatchSnapshot();
  });

  test("if (inp%2 == 0) isEven = true else isEven = false", async () => {
    const exp = `
      if (inp%2 ==0) 
        isEven = true
      else 
        isEven = false
    `;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteExpVisitor();
    const ctx = new RuntimeContext();
    ctx.mem["inp"] = "22";
    runtime.run(astTree, ctx);
    expect(runtime.ctx.mem["isEven"]).toBe(true);
    expect(runtime).toMatchSnapshot();
  });
});
