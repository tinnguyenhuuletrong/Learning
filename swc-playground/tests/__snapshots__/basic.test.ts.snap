// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TLite array res = [1,2,inp > 18,'name'] 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 1",
      "got num: 2",
      "got num: 18",
      "exec > for v1:22, v2:18 -> true",
      "got str: name",
      "got array: [1,2,true,\\"name\\"]",
      "assign res = [1,2,true,\\"name\\"]",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": "22",
      "res": Array [
        1,
        2,
        true,
        "name",
      ],
    },
    "stack": Array [],
  },
}
`;

exports[`TLite arrow function call f = i => { return i + 1}; res = f(1) * f(5) 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got arrow function",
      "assign f = undefined",
      "got num: 1",
      "push stack {\\"i\\":1}",
      "got num: 1",
      "exec + for v1:1, v2:1 -> 2",
      "got block return: 2",
      "pop stack",
      "exec function call f() : args [1] -> res 2",
      "got num: 5",
      "push stack {\\"i\\":5}",
      "got num: 1",
      "exec + for v1:5, v2:1 -> 6",
      "got block return: 6",
      "pop stack",
      "exec function call f() : args [5] -> res 6",
      "exec * for v1:2, v2:6 -> 12",
      "assign res = 12",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "f": [Function],
      "res": 12,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite arrow function call f = i => i + 1; res = f(1) * f(5) 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got arrow function",
      "assign f = undefined",
      "got num: 1",
      "push stack {\\"i\\":1}",
      "got num: 1",
      "exec + for v1:1, v2:1 -> 2",
      "pop stack",
      "exec function call f() : args [1] -> res 2",
      "got num: 5",
      "push stack {\\"i\\":5}",
      "got num: 1",
      "exec + for v1:5, v2:1 -> 6",
      "pop stack",
      "exec function call f() : args [5] -> res 6",
      "exec * for v1:2, v2:6 -> 12",
      "assign res = 12",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "f": [Function],
      "res": 12,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite arrow function call res = map([1,2,3], itm => itm ** 2) 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 1",
      "got num: 2",
      "got num: 3",
      "got array: [1,2,3]",
      "got arrow function",
      "push stack {\\"itm\\":1}",
      "got num: 2",
      "exec ** for v1:1, v2:2 -> 1",
      "pop stack",
      "push stack {\\"itm\\":2}",
      "got num: 2",
      "exec ** for v1:2, v2:2 -> 4",
      "pop stack",
      "push stack {\\"itm\\":3}",
      "got num: 2",
      "exec ** for v1:3, v2:2 -> 9",
      "pop stack",
      "exec function call map() : args [[1,2,3],null] -> res 1,4,9",
      "assign res = [1,4,9]",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {
      "map": [Function],
    },
    "mem": Object {
      "res": Array [
        1,
        4,
        9,
      ],
    },
    "stack": Array [],
  },
}
`;

exports[`TLite condition if (3%2 == 0) isEven = true else isEven = false 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "exec if_branch test",
      "got num: 3",
      "got num: 2",
      "exec % for v1:3, v2:2 -> 1",
      "got num: 0",
      "exec == for v1:1, v2:0 -> false",
      "got bool: false",
      "assign isEven = false",
      "exec if_branch: false",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "isEven": false,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite condition if (inp%2 == 0) isEven = true else isEven = false 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "exec if_branch test",
      "got num: 2",
      "exec % for v1:22, v2:2 -> 0",
      "got num: 0",
      "exec == for v1:0, v2:0 -> true",
      "got bool: true",
      "assign isEven = true",
      "exec if_branch: true",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": "22",
      "isEven": true,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite condition res = 2 > 18 ? 'mature' : 'teen' 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "exec if_branch_expression test",
      "got num: 2",
      "got num: 18",
      "exec > for v1:2, v2:18 -> false",
      "got str: teen",
      "exec if_branch_expression: false -> teen",
      "assign res = \\"teen\\"",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "res": "teen",
    },
    "stack": Array [],
  },
}
`;

exports[`TLite condition res = inp > 18 ? 'mature' : 'teen' 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "exec if_branch_expression test",
      "got num: 18",
      "exec > for v1:22, v2:18 -> true",
      "got str: mature",
      "exec if_branch_expression: true -> mature",
      "assign res = \\"mature\\"",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": "22",
      "res": "mature",
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression 1 + (3 * 5)**2 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 1",
      "got num: 3",
      "got num: 5",
      "exec * for v1:3, v2:5 -> 15",
      "got num: 2",
      "exec ** for v1:15, v2:2 -> 225",
      "exec + for v1:1, v2:225 -> 226",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {},
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = !(+inp < 2 || +inp > 5) 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "exec + for v:-1 -> -1",
      "got num: 2",
      "exec < for v1:-1, v2:2 -> true",
      "exec + for v:-1 -> -1",
      "got num: 5",
      "exec > for v1:-1, v2:5 -> false",
      "exec || for v1:true, v2:false -> true",
      "exec ! for v:true -> false",
      "assign res = false",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": "-1",
      "res": false,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = 'hello' 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got str: hello",
      "assign res = \\"hello\\"",
      "got str: hi",
      "assign res = \\"hi\\"",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "res": "hi",
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = 1 + (3 * 5)**2 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 1",
      "got num: 3",
      "got num: 5",
      "exec * for v1:3, v2:5 -> 15",
      "got num: 2",
      "exec / for v1:15, v2:2 -> 7.5",
      "exec + for v1:1, v2:7.5 -> 8.5",
      "assign res = 8.5",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "res": 8.5,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = inp < 2 || inp > 5 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 2",
      "exec < for v1:-1, v2:2 -> true",
      "got num: 5",
      "exec > for v1:-1, v2:5 -> false",
      "exec || for v1:true, v2:false -> true",
      "assign res = true",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": -1,
      "res": true,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = inp <= 2 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 2",
      "exec <= for v1:2, v2:2 -> true",
      "assign res = true",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": 2,
      "res": true,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = inp > 2 && inp <= 5 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 2",
      "exec > for v1:10, v2:2 -> true",
      "got num: 5",
      "exec <= for v1:10, v2:5 -> false",
      "exec && for v1:true, v2:false -> false",
      "assign res = false",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": 10,
      "res": false,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = inp > 2 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 2",
      "exec > for v1:3, v2:2 -> true",
      "assign res = true",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": 3,
      "res": true,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = inp*=2 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 2",
      "assign inp *= 2",
      "assign res = -2",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": -2,
      "res": -2,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = inp+=2 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 2",
      "assign inp += 2",
      "assign res = 4",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": 4,
      "res": 4,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = inp/=2 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 2",
      "assign inp /= 2",
      "assign res = 1",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": 1,
      "res": 1,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite expression res = inp-=2 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 2",
      "assign inp -= 2",
      "assign res = 0",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": 0,
      "res": 0,
    },
    "stack": Array [],
  },
}
`;

exports[`TLite function call res = res = filter([1, 2, 3, inp], isOdd) 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got num: 1",
      "got num: 2",
      "got num: 3",
      "got array: [1,2,3,21]",
      "exec function call filter() : args [[1,2,3,21],null] -> res 1,3,21",
      "assign res = [1,3,21]",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {
      "filter": [Function],
      "isOdd": [Function],
    },
    "mem": Object {
      "inp": 21,
      "res": Array [
        1,
        3,
        21,
      ],
    },
    "stack": Array [],
  },
}
`;

exports[`TLite function call res = toUpper(toString(inp) + '_a') 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "exec function call toString() : args [22] -> res 22",
      "got str: _a",
      "exec + for v1:22, v2:_a -> 22_a",
      "exec function call toUpper() : args [\\"22_a\\"] -> res 22_A",
      "assign res = \\"22_A\\"",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {
      "toString": [Function],
      "toUpper": [Function],
    },
    "mem": Object {
      "inp": 22,
      "res": "22_A",
    },
    "stack": Array [],
  },
}
`;

exports[`TLite nestest mixed object mixed obj type and dynamic props 1 1`] = `
TLiteExpVisitor {
  "ctx": RuntimeContext {
    "_logs": Array [
      "got str: mr.A",
      "got kv pair: [name,mr.A]",
      "got kv pair: [age,22]",
      "got object: {\\"name\\":\\"mr.A\\",\\"age\\":22}",
      "assign res = {\\"name\\":\\"mr.A\\",\\"age\\":22}",
      "got num: 18",
      "exec > for v1:22, v2:18 -> true",
      "assign res.isMature = true",
      "got str: 123 street A",
      "assign res.house.address = \\"123 street A\\"",
      "got str: table",
      "got str: tv",
      "got str: air-conditioner",
      "got array: [\\"table\\",\\"tv\\",\\"air-conditioner\\"]",
      "assign res.house.items = [\\"table\\",\\"tv\\",\\"air-conditioner\\"]",
    ],
    "_opts": Object {
      "debugTrace": true,
    },
    "funcDb": Object {},
    "mem": Object {
      "inp": 22,
      "res": Object {
        "age": 22,
        "house": Object {
          "address": "123 street A",
          "items": Array [
            "table",
            "tv",
            "air-conditioner",
          ],
        },
        "isMature": true,
        "name": "mr.A",
      },
    },
    "stack": Array [],
  },
}
`;
